src/lang/frontend_ast_lower.rs

```diff
@@ -0,0 +1,798 @@
+        }
+
+        // Convert to &str slices for builder API
+        let public_args_str: Vec<&str> = public_arg_names.iter().map(|s| s.as_str()).collect();
+        let private_args_str: Vec<&str> = private_arg_names.iter().map(|s| s.as_str()).collect();
+
+        // Add predicate to batch using builder
+        let conjunction = pred_def.conjunction_type == ConjunctionType::And;
+
+        if conjunction {
+            cpb_builder
+                .predicate_and(
+                    &name,
+                    &public_args_str,
+                    &private_args_str,
+                    &statement_builders,
+                )
+                .map_err(|e| match e {
+                    crate::frontend::Error::Middleware(mw_err) => LoweringError::Middleware(mw_err),
+                    _ => LoweringError::InvalidArgumentType,
+                })?;
+        } else {
+            cpb_builder
+                .predicate_or(
+                    &name,
+                    &public_args_str,
+                    &private_args_str,
+                    &statement_builders,
+                )
+                .map_err(|e| match e {
+                    crate::frontend::Error::Middleware(mw_err) => LoweringError::Middleware(mw_err),
+                    _ => LoweringError::InvalidArgumentType,
+                })?;
+        }
+
+        Ok(())
+    }
+
+    fn lower_statement_to_builder(
+        &self,
+        stmt: &StatementTmpl,
+    ) -> Result<StatementTmplBuilder, LoweringError> {
+        // Get predicate
+        let pred_name = &stmt.predicate.name;
+        let symbols = self.validated.symbols();
+
+        // Check for native predicates first
+        let predicate = if let Some(native) = native_predicate_from_string(pred_name) {
+            Predicate::Native(native)
+        } else if let Some(&index) = self.batch_predicate_index.get(pred_name) {
+            // References to other predicates in the same batch (including split chains)
+            Predicate::BatchSelf(index)
+        } else if let Some(info) = symbols.predicates.get(pred_name) {
+            match &info.kind {
+                PredicateKind::Native(np) => Predicate::Native(*np),
+                PredicateKind::Custom { index } => Predicate::BatchSelf(*index),
+                PredicateKind::BatchImported { batch, index } => {
+                    Predicate::Custom(middleware::CustomPredicateRef::new(batch.clone(), *index))
+                }
+                PredicateKind::IntroImported {
+                    name,
+                    args_len,
+                    verifier_data_hash,
+                } => Predicate::Intro(IntroPredicateRef {
+                    name: name.clone(),
+                    args_len: *args_len,
+                    verifier_data_hash: *verifier_data_hash,
+                }),
+            }
+        } else {
+            return Err(LoweringError::PredicateNotFound {
+                name: pred_name.clone(),
+            });
```

Comment:
Shouldn't this case be impossible since we already validated the ast?  If that's the case, I think an `unreachable!` would be more appropriate here.